plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java

```diff
@@ -218,17 +225,22 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
                         handle.getConstraintColumns().stream())
                 .collect(toImmutableSet());
 
+        // Create an intermediate HudiTableHandle by applying TupleDomain-based changes
         HudiTableHandle newHudiTableHandle = handle.applyPredicates(
                 newConstraintColumns,
                 partitionColumnPredicates,
-                regularColumnPredicates);
+                regularColumnPredicates,
+                expressionIndexCandidates);
 
         if (handle.getPartitionPredicates().equals(newHudiTableHandle.getPartitionPredicates())
                 && handle.getRegularPredicates().equals(newHudiTableHandle.getRegularPredicates())
                 && handle.getConstraintColumns().equals(newHudiTableHandle.getConstraintColumns())) {
             return Optional.empty();
         }
 
+        // The TupleDomain returned here tells Trino which parts of constraint.getSummary() it still needs to evaluate because the connector might not have fully handled them.
+        // The Expression returned tells Trino which parts of constraint.getExpression() it still needs to evaluate.
+        // The boolean `false` indicates that `constraint.getExpression()` was not fully "absorbed" by the connector's processing and Trino should re-evaluate it.
```

Comment:
So basically, all the predicates or expressions are still kept and re-evaluated by Trino, correct?

Reply:
Yes